tools/testbin

```diff
@@ -0,0 +1,250 @@
+check_control() {
+    count=`grep -c "$UPGRADEABLE_RE" $CONTROL || :`
+    if [ "$count" -ne 1 ]; then
+        die "$CONTROL must contain exactly one line matching /$UPGRADEABLE_RE/"
+    fi
+}
+
+# Requires:
+# - PG_PORT_BASE
+# - PG_VERSION
+# Sets:
+# - DB
+# - PG_PORT
+select_pg() {
+    PG_PORT=$(( $PG_PORT_BASE + $PG_VERSION ))
+    DB=$DB_ROOT/$PG_VERSION
+}
+
+# Start postgres and create a database.
+# Must select_pg first.
+# Sets:
+# - PG14PID (or PG13PID etc.; depends on value of PG_VERSION)
+start_postgres() {
+    $nop initdb "$DB"
+    $nop mkdir "$DB/lock"
+    $nop postgres -D "$DB" -k "$DB/lock" -p "$PG_PORT" & eval PG${PG_VERSION}PID=$!
+    for i in 1 2 3 4 5; do
+        if $nop psql -h 127.1 -p $PG_PORT postgres < /dev/null; then
+            $nop createdb -h 127.1 -p $PG_PORT
+            return
+        fi
+        echo $i...
+        sleep 1
+    done
+    die "failed to start postgres"
+}
+
+# Stop postgres and create a database.
+# Use select_pg to set which one to stop.
+stop_postgres() {
+    eval pid=\$PG${PG_VERSION}PID
+    [ -n "$pid" ] && $nop kill $pid
+    $nop rm -rf "$DB"
+}
+
+start_test() {
+    start_postgres
+    $nop cargo run --manifest-path tools/update-tester/Cargo.toml -- create-test-objects -u $LOGNAME -h 127.1 -p $PG_PORT
+}
+
+finish_test() {
+    $nop cargo run --manifest-path tools/update-tester/Cargo.toml -- validate-test-objects -u $LOGNAME -h 127.1 -p $PG_PORT
+    stop_postgres
+}
+
+test_deb() {
+    ARCH=`dpkg --print-architecture`
+    # Could parameterize these, but we're only supporting Ubuntu 20.04 right now.
+    OS_NAME=ubuntu
+    OS_VERSION=20.04
+    EPOCH=
+    MIN_DEB_ARM=`cmp_version $MIN_DEB_ARM`
+    MIN_DEB_EPOCH=`cmp_version $MIN_DEB_EPOCH`
+    for from_version; do
+        # We released 1.10.0-dev by accident.  We have to support upgrades
+        # from it (and we tested that at the time), but we pulled the deb, so
+        # we can't test it here.
+        [ $from_version = 1.10.0-dev ] && continue
+        cmp_version=`cmp_version $from_version`
+        [ "$ARCH" -eq arm64 ] && [ $cmp_version -lt $MIN_DEB_ARM ] && continue
+        [ $cmp_version -ge $MIN_DEB_EPOCH ] && EPOCH=1:
+        for PG_VERSION in $PG_VERSIONS; do
+            select_pg $PG_VERSION
+            deb=timescaledb-toolkit-postgresql-${PG_VERSION}=${EPOCH}${from_version}~${OS_NAME}${OS_VERSION}
+            $nop sudo apt-get -qq install $deb
+
+            PATH=/usr/lib/postgresql/$PG_VERSION/bin:$PATH start_test
+        done
+        for PG_VERSION in $PG_VERSIONS; do
+            select_pg $PG_VERSION
+            deb=timescaledb-toolkit-postgresql-${PG_VERSION}_${TOOLKIT_VERSION}~${OS_NAME}${OS_VERSION}_${ARCH}.deb
+            $nop sudo dpkg -i "$BINDIR/$deb"
+
+            finish_test
+
+            $nop sudo dpkg -P timescaledb-toolkit-postgresql-$PG_VERSION
+        done
+    done
+}
+
+test_rpm() {
```

Comment:
The rpm build was already broken when I wrote the first draft of this, so this has never been tested and it has rotted as the deb version was refined.  I'll return to this after fixing the rpm build.  That means rpm testing may need to be manual again for this release.  But we're getting close!